### DISCUSSION TITLE
Should accessor methods be available through Column::make?

### DISCUSSION THREAD

**Discussion by Edim7** [rappasoft/laravel-livewire-tables#1103] [Q&A] [answered]
Hi everyone 👋 

My question is about the Column::make functionality and how it interfaces with the Model to retrieve data when we are not targeting a specific attribute from the table, rather an attribute that is retrieved through an accessor method.

According to this [discussion](https://github.com/rappasoft/laravel-livewire-tables/discussions/273) this should work, however I'm having an issue with the missing column error. (SQLSTATE[42S22]: Column not found: 1054 Unknown column 'users.full_name' ...)

In my example I have a **users** table in the db that has **first_name** and **last_name** attributes.
Within the model I have a getter for full name through an accessor method named `getFullNameAttribute()`

accessor method:
```
 public function getFullNameAttribute()
    {
        return $this->first_name.' '.$this->last_name;
    }
```
In the the UserTable the column looks like this:

```
Column::make(__('Full Name'), 'full_name')
                ->sortable()->searchable()
```

I have a feeling this should work, and I don't understand why it isn't working so should this work, and do you have any advice on how would I go about understanding on why it doesn't?

Update:

I found [this discussion](https://github.com/rappasoft/laravel-livewire-tables/discussions/831) as well noting that accessor methods worked in v1, but don't work in v2 so clear cut answer would be helpful to decide if this is supported, and if not maybe to open up a ticket for it to be re-introduced.

If anyone is looking for a workaround querying on the model works with the format() or label() option on the column.

**Comment by lrljoe** [comment, marked as answer]
So!  The accessor methods won't work as those all feed into the Builder query, and so they need to be actual database fields.

A few different options here.

I've got a PR open #1082 - which will allow you to add some selectRaw columns in there, so you'll be able to do a concat easily enough and return it as a field for use.  Not sure if/when that'll make it into core though!

Another alternative of course depending on your database software is to define the attribute in there

Or you can of course as you said create a Label column.

**Comment by Edim7** [comment]
As of now I'll use a solution inspired by [imabug](https://github.com/rappasoft/laravel-livewire-tables/discussions/831#discussioncomment-3717039) and I will consider this to be by design for the v2 version, which as you stated **requires actual database fields**.

In the meantime if anyone is looking for a _clean_ solution I suggest using a `->format()` which lets you get the accessor method attributes in the table they are located in, as well as in cases where you are jumping through relationships to get that accessor method attribute.

This would be my recommended way of getting to the accessor methods in v2:

```
Column::make('Full name in Users table', 'id')
                ->format(
                    fn($value) => User::find($value)->full_name
                ->sortable()->searchable(),
```


```
Column::make('Full name from Users somewhere else', 'user_fk_attrib')
                ->format(
                    fn($value) => User::find($value)->full_name
                )->sortable()->searchable(),
```

**Reply by lrljoe** [reply]
Quick caveat on your approach, it will not play nicely with column select if you use a field that is used elsewhere.

If you don't need searchable/sortable then use a Label.
If you do then chuck a different field at it to get it working, and customise the sortable/searchable callback

I've got a fix pending for one Column Select issue, and I'll put another together to address this issue once @rappasoft has caught up with my dozen or so PRs of mixed value amongst others contributing!

Sadly if I keep creating PRs then I inevitably mess up my branches and end up spending a day unpicking them all.

**Comment by lrljoe** [comment]
@Edim7  - another quick point, if you use the Builder function then you can retrieve the relevant attributes, and you won't need to do the additional model load, not to mention that you'll end up doing it 100 times if you've got 100 rows to a page, which will quickly slow things down.

If you create a new discussion/issue with your component, then I'll give you a few pointers on how to do it efficiently where I can.

Don't forget that with format() you have the following:
```php
    ->format(
        fn($value, $row, Column $column) => $row->first_name . ' ' . $row->last_name
    ),
```

If you use with() in your builder method, then you can access the attribute via $row->attributeName, and it removes the necessity to do your lookup.

**Reply by Edim7** [reply]
Thank you for the help and useful explanations, really appreciate it
